modules/kinesis/src/main/resources/application.conf

```diff
@@ -0,0 +1,55 @@
+{
```

Comment:
How many collector payloads do you think it's possible to be holding in memory at the same time with these default settings?  If you're processing, say, 5 kinesis shards per enrich instance.

I count 190,000 (admittedly a worse case scenario).  And a further 10,000 more per additional shard.

Here's how:

Each polling batch is [10000 records](https://github.com/snowplow/enrich/blob/5778418d5ccf8ace73844cf5ed0110b768e3407c/modules/kinesis/src/main/resources/application.conf#L8).  The fs2-aws library uses a buffer that holds [10 batches](https://github.com/laserdisc-io/fs2-aws/blob/d07461a0d2515688bd4f18f87e8c14b80acd8773/fs2-aws/src/main/scala/fs2/aws/kinesis/KinesisSettings.scala#L43).  Plus there can be extra batch _per shard_ [on this line waiting to enqueue](https://github.com/laserdisc-io/fs2-aws/blob/v3.1.1/fs2-aws/src/main/scala/fs2/aws/kinesis/Kinesis.scala#L86).

Then you have the batch you're currently processing, plus three extra batches because of the prefetches [here](https://github.com/snowplow/enrich/blob/5778418d5ccf8ace73844cf5ed0110b768e3407c/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala#L81), [here](https://github.com/snowplow/enrich/blob/5778418d5ccf8ace73844cf5ed0110b768e3407c/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala#L90) and [here](https://github.com/snowplow/enrich/blob/5778418d5ccf8ace73844cf5ed0110b768e3407c/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala#L100)

This is just to explain why I've been questioning memory problems, especially for large collector payloads.  Understanding the scope of this problem will be difficult, because we would need to explore regular processing, and also error-handling scenarios, e.g. kinesis problems, enrichment problems.  The latter are harder to predict or simulate.

Reply:
Very interesting!

I agree with your way of counting. There has been only a little change, I removed the `.prefetch` to now use `>1` for `sink.concurrency` [here](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala#L93) so now it's `concurrency - 1` additional shards instead of three.

Because of the enqueuing for each shard I'm afraid the memory used by the app is unbounded. So we need to make sure that our auto-scaling is good enough so that new instances are spawned before the app gets a chance to go OutOfMemory. I'll look at our scaling strategy in Terraform. But I think that whatever the strategy, it would be great to use the average record size of a customer to determine the memory allocated to the JVM. /cc @oguzhanunlu @jbeemster